### ISSUE TITLE
How to tell if $watch events are local or remote

### ISSUE THREAD

**Issue by jaymanned** [firebase/angularfire#713]
When I call `$add` on a local `$firebaseArray`, a new record is added both to the local object and eventually pushed to the server. Since the addition is made locally I would not expect to receive a `child_added` event when using `$watch`.

However I see at #368 that this behaviour is intentional. So the question is how can I tell at `$watch` if the change was triggered remotely or locally? In my case I use `$watch` only to monitor the remote changes.

**Comment by katowulf** [comment]
Without your detailed use case and constraints, it's going to be hard to offer specifics. This is likely an [XY problem](http://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378#66378).

Generally speaking, the correct answer here is to treat local and remote events the same way. Use a one-directional model as it's much simpler (a la React and Angular 2 change detection) and easier to predict. 

More generally speaking, whenever you find yourself using `$loaded` for anything other than debugging, or `$watch` for anything other than triggering sorting or some sort of event emitting, then you should probably be using `$extend` instead.

**Comment by jaymanned** [comment]
Hi Kato

As you might remember, originally `$watch` prior to #368 was triggered only by remote changes. Then after that request, you guys added local events to the mix. That means that during a period of time, the remote only approach made sense for you guys. So probably we can cross out the XY problem, don't you think. 

In simpler apps where objects read from FB can be binded directly into the view, watching both local and remote events and processing them identically would make good sense. But for more complex apps with highly denormalized data structures, where the model objects have to be assembled using data from multiple FB locations, detecting just the remote changes is convenient, since there is no need to reassemble objects on local changes. Regretfully my app is a bit too complex to provide an intuitive enough use case for us to discuss.

In short, I suggest that `$watch` itself informs on the nature of the event.

**Comment by katowulf** [comment]
Thanks for the feedback! It sounds like the best solution would be to $extend the $watch method and make it behave however you like then.

**Comment by jaymanned** [comment]
But even if I `$extend` the `$watch` method, I need a FB way to detect local vs remote, don't I?

**Comment by katowulf** [comment]
I think the answer is pretty much the change list you linked. If it comes from $save, it's local. An alternative would be to keep a clean copy of the latest server data (sent by $$updated in the case of $firebaseObject) and compare whenever $watch is called. Either seems like it should work for your needs.

**Comment by jaymanned** [comment]
Thanks Kato, I'll give it a try. But please keep in mind my request: make
$watch flag if events are remote or not.

On Tue, 10 May 2016 at 17:45 katowulf <notifications@github.com> wrote:

> I think the answer is pretty much the change list you linked. If it comes
> from $save, it's local. An alternative would be to keep a clean copy of the
> latest server data (sent by $$updated in the case of $firebaseObject) and
> compare whenever $watch is called. Either seems like it should work for
> your needs.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/firebase/angularfire/issues/713#issuecomment-218200288>
>
-- 
Javier Aragonés
javier.aragones@gmail.com

**Comment by katowulf** [comment]
Thanks, Javier.  I really appreciate the feedback and definitely hear that you'd like to be able to tell server-triggered events from local. Please continue to provide feedback in the future as our community has been instrumental in shaping Firebase. Solid and detailed use cases are always appreciated as well.

Firebase uses a different paradigm, and thinking local vs. remote often leads to issues with connectivity and offline usage--things Firebase handles very gracefully, in part because we enforce this model of one-directional event streams and the complexities that come from trying to manage local vs remote separately.

While I'm not willing to invest any time on this, considering that this goes against the paradigm of the underlying SDK, and considering that it can be resolved by extending the lib to your needs. I do hear you that it's a pain point for your use case. 

Ping me back here if you have issues getting things customized to your needs. Glad to help.

**Comment by jaymanned** [comment]
Hi Kato

Thanks again for your support (I say again because in the last year I've read sooo many FB issues solved by you that have be invaluable to me).

This thing about one-directional event streams, is a very interesting topic and I would love to learn more about this paradigm. Have you found any article dealing about it that you considered worthwhile?

**Comment by katowulf** [comment]
Check out the section on _The Flux solution_, particularly the graph and description of _Command Query Responsibility Segregation_, in this article: https://medium.com/swlh/the-case-for-flux-379b7d1982c6#.smho7l8b6
 
Discussion of different unidirectional data flows, including how Firebase fits into this paradigm: https://medium.com/@AdamRNeary/unidirectional-data-flow-yes-flux-i-am-not-so-sure-b4acf988196c#.d0om899yx

The diagrams under Structure and Data flow in Flux's overview explain unidirectional data flow in a great visual form: https://facebook.github.io/flux/docs/overview.html

Why does React emphasize unidirectional data flow?
https://hashnode.com/post/why-does-react-emphasize-on-unidirectional-data-flow-and-flux-architecture-ciibz8ej600n2j3xtxgc0n1f0

Redux's Data Flow article
http://redux.js.org/docs/basics/DataFlow.html

The key to applying these concepts to Firebase is captured in the graphic from the Flux Overview pasted just below. In the case of Firebase, the Dispatcher is the Firebase SDK (or a thin wrapper on top of it like AngularFire). The controllers/app logic direct Actions back to the dispatcher rather than directly updating models.

![image](https://cloud.githubusercontent.com/assets/959972/15160574/224e5a8c-16af-11e6-952b-e35834f5dfc3.png)

I hope that helps!

**Comment by jaymanned** [comment]
Wonderful Kato. I started reading yesterday night. It might be exactly what I need. Cheers

**Comment by jaymanned** [comment]
Hi Kato

I've been checking Angularfire's code and haven't been able to find a way to tell local from remote changes because at the end $watch listeners are invoked via plain firebase .on() method events which also do not distinguish local changes from remote. 

From some issues I read on SO, I think I found a practical way to tell local from remote. It seems local events would fire back immediately (i.e. synchronously) because they do not need to propagate first to the firebase servers. However remote changes fire asynchronously as they need to come from the servers. If this is correct, it would be possible to flag internally local changes to the listeners so they could ignore these local events.

However I haven't been able to find anything in the Docs to support this statement. Kato, is this assumption correct?

**Comment by jaymanned** [comment]
In case any one is interested, found [this article](https://firebase.googleblog.com/2014/05/handling-synchronized-arrays-with-real.html) by @katowulf confirming that local events fire back immediately.
